### DISCUSSION TITLE
How to run it synology task planer

### DISCUSSION THREAD

**Discussion by git-newbe** [icloud-photos-downloader/icloud_photos_downloader#546] [Q&A]
Hi, 
I'm an absolute new-be to all of this.
I already managed to install the docker and run the following statement successfully in the termial.

icloudpd --directory /icloud/media \--folder-structure {:%Y/%Y-%m} \--username 'myuser@icloud.com' \--password 'password' \--recent 500 \--auto-delete

so far so good.

Now I would like to run it from time to time via a sh script using the task planner of my Synology NAS.

I created an sh file containing (similar as advised in readme file)

1.)docker pull icloudpd/icloudpd
2.)docker run -it --rm --name icloudsync \
3.)icloudpd/icloudpd:latest \
4.)icloudpd --directory /icloud/media \--folder-structure {:%Y/%Y-%m} \--username 'myuser@icloud.com' \--password 'password' \--recent 500 \--auto-delete

Now I'm getting the following error:

invalid reference format
the input device is not a TTY
/volume1/docker/icloud/Copy.sh: line 3: icloudpd/icloudpd:latest: No such file or directory
/volume1/docker/icloud/Copy.sh: line 4: icloudpd: command not found

What do i need to do?; 
Will it automatically create a new docker container? Or open my existing one, called icloudsync (icloudpd/icloudpd:latest) 
Or need the docker always run?

Thanks for a detailed guidance ;)

Max

**Comment by AndreyNikiforov** [comment]
> 4.)icloudpd --directory /icloud/media ...

Since you are running icloudpd in docker, then your command will be `docker run -it --rm --name icloudsync icloudpd/icloudpd:1.8.0 icloudpd --directory /icloud/media ...`

Note that docker is like a virtual machine with its own filesystem, so you have to map your local (==host) file system to internal docker filesystem: `docker run -it --rm --name icloudsync -v /my/host/Photos:/icloud/media icloudpd/icloudpd:1.8.0 icloudpd --directory /icloud/media ...`

Just to decrypt docker commands above: `docker CMD DOCKER_PARAMS IMAGE:VERSION CMD_TO_RUN_INSIDE PARAMS_FOR_CMD_INSIDE` our `IMAGE:VERSION` pair is `icloudpd/icloudpd:1.8.0`

**Comment by swxxii** [comment]
I got the docker image running on my Synology as a scheduled task which every 2 hours:
- boots up a new docker container
- runs icloudpd to sync photos
- terminates, destroying the container (we don't need it anymore).

First pull the image with Docker UI on Synology. (I'll assume the reader knows how to do this.)

Then create a script `icloudpd.sh` somewhere which looks something like this (sub your own folders). 
```
docker run --rm --name icloudpd \
    -v /volume1/Media/backup/iCloudPhotos:/data \
    -v /volume1/configs/icloudpd/cookies:/cookies \
    -e TZ=Australia/Melbourne \
    icloudpd/icloudpd:latest \
    icloudpd --directory /data \
    --cookie-directory /cookies \
    --folder-structure {:%Y/%Y-%m-%d} \
    --username 'xxxxxx' \
    --password 'xxxxxx' \
    --recent 500 \
    --size original \
    --auto-delete
```
Then make script executable with `chmod a+x /path/to/icloudpd.sh`.

Then SSH into your NAS and run the script once to make sure it works. 

`sudo /path/to/icloudpd.sh`

If prompted for iCloud 2FA proceed but if you mounted cookies path properly it should run without prompting next time.

Go to Synology settings and create the scheduled task to run whenever you would like. 

In the "User defined script" box type `bash /volume1/configs/scripts/icloudpd.sh` or similar for your own setup.

Notes:
- The error `the input device is not a TTY` was solved by removing the `-it` (interactive) flag from the command.
- I could not get the keychain working so yes, my password is in plain text in the script. This is not ideal. Advice welcome!
- I have to run scheduled task as root so it can connect to the docker daemon. If there is a better way to do this, please let me know.

**Comment by git-newbe** [comment]
Hi,

I do it meanwhile like this.

1.) Set up a "permanent docker" with latest version (or version I want). Run this docker "manual" and setup cookies and keyring.

2.) I schedule in synlogy task planer only:

docker start icloud 
docker exec icloud sh /icloud/icloud_sync.sh
docker stop icloud

In the script i use the paramter i want.

From my point of view it's not needed to always check for new docker release and download. Even if a new version is available you might need to adjust the parameter. From time to time I check what new features have been deployed and then i setup an new "permanent docker"

**Reply by swxxii** [reply]
It’s funny you mention because I just changed to a similar method  after realising i could call it from a script that sits outside the container ^_^

**Comment by AndreyNikiforov** [comment]
current 1.23.4 can run as [headless daemon](https://icloud-photos-downloader.github.io/icloud_photos_downloader/webui.html) (`--watch-withinterval 3600 --mfa-provider webui --password-provider webui`)
